### ISSUE TITLE
defacing non T1 image eg FLASH

### ISSUE THREAD

**Issue by agramfort** [mne-tools/mne-bids#650]
I've being struggling for way too long to be able to write my FLASH sequence with the same defacing at the T1.

Here is my attempt:

```
import os.path as op

import numpy as np
import matplotlib.pyplot as plt

import mne
from mne.datasets import sample
from nilearn.plotting import plot_anat

from mne.source_space import head_to_mri

from mne_bids import BIDSPath, write_anat

data_path = sample.data_path()
raw_fname = op.join(data_path, 'MEG', 'sample', 'sample_audvis_raw.fif')
events_data = op.join(data_path, 'MEG', 'sample', 'sample_audvis_raw-eve.fif')
output_path = op.join(data_path, '..', 'MNE-sample-data-bids-debug')

raw = mne.io.read_raw_fif(raw_fname)
subject = '01'

###############################################################################
# Get Landmarks from MEG file, 0, 1, and 2 correspond to LPA, NAS, RPA
# and the 'r' key will provide us with the xyz coordinates
pos = np.asarray((raw.info['dig'][0]['r'],
                  raw.info['dig'][1]['r'],
                  raw.info['dig'][2]['r']))

# Load the transformation matrix and show what it looks like
trans_fname = op.join(data_path, 'MEG', 'sample',
                      'sample_audvis_raw-trans.fif')
trans = mne.read_trans(trans_fname)

t1_pos = head_to_mri(pos=pos,
                     subject='sample',
                     mri_head_t=trans,
                     subjects_dir=op.join(data_path, 'subjects'))
mri_pos = t1_pos / 1e3  # put landmarks in meters

montage = mne.channels.make_dig_montage(
    lpa=mri_pos[0],
    nasion=mri_pos[1],
    rpa=mri_pos[2],
    coord_frame='mri'
)

###############################################################################
# FLASH
flash_mgh_fname = \
    op.join(data_path, 'subjects', 'sample', 'mri', 'flash', 'mef05.mgz')

flash_bids_path = \
    BIDSPath(subject=subject, root=output_path, suffix='FLASH')

anat_path = write_anat(
    image=flash_mgh_fname,
    bids_path=flash_bids_path,
    landmarks=montage,
    deface=True,
    overwrite=True,
    verbose=True
)

# Plot it
fig, axs = plt.subplots(3, 1)
for point_idx, label in enumerate(('LPA', 'NAS', 'RPA')):
    plot_anat(str(anat_path.fpath), axes=axs[point_idx],
              cut_coords=mri_pos[point_idx, :] * 1e3,
              title=label, vmax=800)
plt.show()
```

any help would be really valuable here.

@alexrockhill maybe?

**Comment by agramfort** [comment]
<img width="859" alt="Screenshot 2020-12-03 at 18 13 17" src="https://user-images.githubusercontent.com/161052/101063583-3f190100-3593-11eb-8cbd-9c0ae6e96ee4.png">

as you can see it is broken....

**Comment by agramfort** [comment]
What is clear is that the T1 and FLASH images do not have the same affine. But MRI RAS coord should be compatible with the flash image. There is really something I don't understand here....

**Comment by alexrockhill** [comment]
I'm taking a look now

**Comment by agramfort** [comment]
the flash image is in the same dataset

>

**Comment by alexrockhill** [comment]
So in freeview the FLASH and T1 are overlaid very nicely but the affine for the FLASH is
```
array([[ -0.02463216,  -0.03026463,  -0.99923837,  97.68000793],
       [ -0.977844  ,   0.20857805,   0.0177875 , 105.76734161],
       [ -0.20788085,  -0.97753739,   0.03473151, 121.11663818],
       [  0.        ,   0.        ,   0.        ,   1.        ]])
```

Whereas the affine for the T1 is:

```
array([[  -1.  ,    0.  ,   -0.  ,  122.73],
       [   0.  ,    0.  ,    1.  , -118.96],
       [   0.  ,   -1.  ,    0.  ,  100.71],
       [   0.  ,    0.  ,    0.  ,    1.  ]])
```

That seems to me what's likely to be causing the weird results. If freesurfer (and nilearn.plot_anat) is correctly overlaying them though, they must be loading it in differently. I'll dig around in `plot_anat` and see if there are any differences in how the image object is loaded.

**Comment by alexrockhill** [comment]
![raw+trans](https://user-images.githubusercontent.com/13473576/101084314-a1a5d900-3562-11eb-9741-675777cf778d.png)

Hmm providing the raw and the trans gives a different but incorrect deface as well so there might be more than one issue...

**Comment by alexrockhill** [comment]
So I've located how they do it in nilearn to `check_niimg` on line 170 here https://github.com/nilearn/nilearn/blob/c10248e43769f37eaea804f64d44a7816e3c6e03/nilearn/_utils/niimg_conversions.py. It is at this point in `plot_anat` that the FLASH image data and the T1 image data match up after initially not matching up after being loaded with `nib.load`.

Hopefully answer shortly, just leaving messages to myself/others in the future.

**Comment by alexrockhill** [comment]
So I think the issue is that the `trans` defines the head to T1 transformation matrix but the T2 in the sample dataset has a different order of dimensions/orientation codes when loaded in with `nib.load`. So the T2 needs to go to T1 space or everything needs to go to a common space such as RAS at the end. It got quite complicated how nilearn did it so I changed back to looking at the implementation in `write_anat`...

**Comment by alexrockhill** [comment]
Ah okay solved it, PR coming...

In this code in `_deface`:

```
    # x: L/R L+, y: S/I I+, z: A/P A+
    t1w_data = t1w.get_fdata().copy()
    idxs_vox = np.meshgrid(np.arange(t1w_data.shape[0]),
                           np.arange(t1w_data.shape[1]),
                           np.arange(t1w_data.shape[2]),
                           indexing='ij')
    idxs_vox = np.array(idxs_vox)  # (3, *t1w_data.shape)
    idxs_vox = np.transpose(idxs_vox,
                            [1, 2, 3, 0])  # (*t1w_data.shape, 3)
    idxs_vox = idxs_vox.reshape(-1, 3)  # (n_voxels, 3)

    mri_landmarks_ras = apply_trans(t1w.affine, mri_landmarks)
    ras_meg_t = \
        get_ras_to_neuromag_trans(*mri_landmarks_ras[[1, 0, 2]])
```

The image is assumed to be in LIA coordinates. It should just be correct to RAS with `img = nib.as_closest_canonical(img)` which should solve the issue.

**Comment by alexrockhill** [comment]
@agramfort, it looks like it's fixed on my deface branch for the example you posted now and it works on the T1

**Comment by hoechenberger** [comment]
This has been fixed via #651
